stonesoup/dataassociator/general.py

```diff
@@ -0,0 +1,179 @@
+from typing import Set, List, Union, Tuple, Dict, Optional
+import numpy as np
+from stonesoup.base import Property, Base
+from stonesoup.measures import GenericMeasure, Measure
+from stonesoup.dataassociator._assignment import assign2D
+from stonesoup.types.association import Association
+from stonesoup.types.track import Track
+from stonesoup.types.state import State
+
+
+class GeneralAssociationGate(Base):
+
+    @abstractmethod
+    def __call__(self, item1, item2) -> bool:
+        raise NotImplementedError
+
+
+class MeasureThresholdGate(GeneralAssociationGate):
+
+    minimise_measure: bool = Property(default=True)
+    association_threshold: float = Property()
+    measure: GenericMeasure = Property()
+
+    def __call__(self, item1, item2) -> bool:
+        distance_measure = self.measure(item1, item2)
+        if self.minimise_measure:
+            return distance_measure <= self.association_threshold
+        else:  # maximise measure
+            return self.association_threshold <= distance_measure
+
+
+class RecentTrackMeasure(GenericMeasure):
+    state_measure: Measure = Property()
+    max_timestep_to_look_back: int = Property(default=10)
+
+    def __call__(self, track1: Track, track2: Track) -> Optional[float]:
+
+        track_1_dict: Dict[datetime, State] = \
+            {state.timestamp: state for i, state in enumerate(reversed(track1.states))
+             if i < self.max_timestep_to_look_back}
+
+        track_2_dict = {state.timestamp: state for i, state in enumerate(reversed(track2.states)) if
+                        i < self.max_timestep_to_look_back}
+
+        all_times = set(track_1_dict.keys()) | set(track_2_dict.keys())
+
+        measures = []
+        for time in all_times:
+            state1 = track_1_dict.get(time)
+            state2 = track_2_dict.get(time)
+
+            if state1 is not None and state2 is not None:
+                measures.append(self.state_measure(state1, state2))
+
+        if len(measures) != 0:
+            return mean(measures)
+        else:
+            return None
+
+
+class OneToOneAssociatorWithGates(Associator):
+
+    gates: List[GeneralAssociationGate] = Property(default=None)
+
+    fail_gate_value: Union[int, float, complex, np.number] = Property(default=None)
+
+    measure: GenericMeasure = Property()
+    association_threshold: float = Property(default=None)
+
+    maximise_measure: bool = Property(default=False)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        if self.gates is None:
+            self.gates = []
+
+        if self.fail_gate_value is None:
+            if self.maximise_measure:
+                self.fail_gate_value = 0
+            else:
+                self.fail_gate_value = 1e6
+
+        if self.association_threshold is None:
+            if self.maximise_measure:
+                self.association_threshold = 0
+            else:
+                self.association_threshold = 1e6
```

Comment:
This `__init__` isn't necessary - Stone Soup handles for you.  If you set the default for `gates` to `[]` and the other two here to `None`, it would be a lot cleaner.  Will just require a few changes/extra if statements later on, but setting `association_threshold` to arbitrarily large/small numbers is probably best avoided